Avoid unnecessary reflection

ADmad opened the pull request.

ADmad committed:
Remove unneeded check.

garas commented:
Abstract/interface controllers will now respond with error 500 instead of 404.

dereuromark commented:
I agree, the 2nd iteration (removing MissingControllerException wrapping) was a bit too much here.

markstory commented on src/Controller/ControllerFactory.php:
```diff
@@ -38,24 +37,19 @@ class ControllerFactory implements ControllerFactoryInterface
         }
-        /** @var string $className */
-        $reflection = new ReflectionClass($className);
-        if ($reflection->isAbstract() || $reflection->isInterface()) {
```
As others mentioned this will cause 500s when the framework used to emit 404, this will generally only impact applications that use fallback routing and abstract/interfaces in their controllers.

Does the usage or reflection have any runtime impact? I know that in early PHP5 Reflection could be a performance problem, but I'm not sure that holds true today.

dereuromark replied:
Controller reflection check is IMO never a performance impact, as this is always only a singular PHP invocation per request lifecycle. I think this is in the nano-area of relevance.

The main idea of this change was probably to have less complicated code. And somewhere in the middle we can probably find a good solution here.

ADmad commented:
I have added try/catch to ensure 404 is returned as before.

dereuromark commented on src/Controller/ControllerFactory.php:
```diff
@@ -38,23 +38,23 @@ class ControllerFactory implements ControllerFactoryInterface
+            /** @psalm-suppress UndefinedClass */
+            $controller = new $className($request);
+        } catch (Error $e) {
             $this->missingController($request);
```
Before, you had a 2nd argument to rethrow $e into the chain. Might be worth resurrecting that change from before.

ADmad replied:
Can't you just pretend you didn't see that commit? :stuck_out_tongue:

garas replied:
`Error` can be thrown not only because controller is not instantiable. Many things can happen in many parts of constructor.

I would stay with reflection. Only maybe simplify to `ReflectionClass::isInstantiable()`.

dereuromark replied:
@garas One could argue that once the controller is instantiable, that anything else that goes wrong warrants an internal error 500, since then there is a real issue.
But `ReflectionClass::isInstantiable()` sounds also like a good idea.

ADmad commented:
I have reverted all changes and just removed the unneeded `isInterface()` check. Since `App::className()` always does `class_exists()` check it can never return a interface name.
